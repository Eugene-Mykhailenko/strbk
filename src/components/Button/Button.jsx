import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import styles from './Button.scss';

import Icon from '../Icon/Icon';

const Button = (props) => {
  const {
    text,
    intent,
    size,
    block,
    isLoading,
    disabled,
    type,
    icon,
    iconSize,
    onClick,
  } = props;

  const getClassNames = () =>
    classNames(styles.btn, styles[`btn-${intent}`], {
      [styles[`btn-${size}`]]: size,
      [styles['btn-block']]: block,
      [styles.disabled]: disabled,
      [styles['is-loading']]: isLoading,
    });

  const renderIcon = () =>
    icon &&
    !isLoading && (
      <Icon
        className={styles['btn-icon']}
        glyph={icon}
        width={iconSize}
        height={iconSize}
      />
    );

  const handleClick = (event) => {
    !disabled && onClick(event);
  };

  return (
    <button
      className={getClassNames(styles)}
      disabled={disabled}
      type={type}
      onClick={handleClick}
    >
      {renderIcon()}
      {text}
    </button>
  );
};

Button.styles = styles;

Button.propTypes = {
  /** Button text. */
  text: PropTypes.string,
  /** Variant of button visualization. */
  intent: PropTypes.string,
  /** Button icon name. */
  icon: PropTypes.string,
  /** Button icon size. */
  iconSize: PropTypes.number,
  /** Render button full width. */
  block: PropTypes.bool,
  /** Button isLoading state. */
  isLoading: PropTypes.bool,
  /** Button type. */
  type: PropTypes.string,
  /** Button size. */
  size: PropTypes.string,
  /** Button disabled state. */
  disabled: PropTypes.bool,
  /** It's an action when the user clicks on the button. */
  onClick: PropTypes.func.isRequired,
};

Button.defaultProps = {
  text: undefined,
  intent: 'main',
  block: false,
  isLoading: false,
  type: 'button',
  iconSize: 20,
  disabled: false,
  icon: null,
  size: undefined,
};

export default Button;
