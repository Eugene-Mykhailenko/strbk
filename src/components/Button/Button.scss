@import '../../assets/styles/abstracts/variables';
@import '../../assets/styles/components/spinner';
// Buttons

.btn {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  height: 48px;
  padding: 0 15px;
  border: none;
  font-size: 14px;
  font-family: inherit;
  text-transform: uppercase;
  text-decoration: none;
  vertical-align: middle;
  font-weight: bold;
  color: inherit;
  background: none;
  cursor: pointer;
  outline: none;

  &::before,
  &::after {
    content: '';
    display: inline-block;
  }

  &-icon {
    display: inline-block;
    vertical-align: middle;
    margin-right: 8px;
  }

  &-main {
    border: 1px solid var(--color_1);
    color: var(--color_3);
    background-color: var(--color_1);

    &.loading {
      &::before {
        background-color: var(--color_9);
      }
    }
  }

  &-primary {
    border: 1px solid var(--color_12);
    color: var(--color_9);
    background-color: var(--color_12);
  }

  &-secondary {
    border: 1px solid var(--color_11);
    color: var(--color_9);
    background-color: var(--color_11);
  }

  &-cancel {
    border: 1px solid var(--color_7);
    color: var(--color_5);
    background-color: var(--color_7);
  }

  &-default {
    border: 1px solid var(--color_7);
    color: var(--color_6);
    background-color: var(--color_9);
  }

  &-simple {
    align-items: center;
    justify-content: center;
    height: 44px;
    padding: 0 10px;
    border: 1px solid transparent;
    color: var(--color_3);
    background-color: transparent;

    .btn-icon {
      margin: 0 5px;
    }
  }

  //size
  &-sm {
    height: 32px;
    padding: 0 20px;
    font-size: 12px;
    font-weight: 400;
    text-transform: none;
  }

  //disabled
  &.disabled,
  &[disabled] {
    border-color: var(--color_13);
    color: var(--color_6);
    background-color: var(--color_13);
    cursor: default;

    &.is-loading {
      &::before {
        background-color: var(--color_1);
      }
    }
  }

  &-block {
    display: flex;
    width: 100%;

    + .btn {
      margin-top: 5px;
    }
  }

  &.is-loading {
    @extend .spinner;

    color: transparent;

    &.btn-main {
      &::before {
        background-color: var(--color_9);
      }
    }
  }
}

.btn-row {
  display: flex;
  flex-wrap: nowrap;
  width: 100%;

  .btn + .btn {
    margin-top: 0;
    margin-left: 5px;
  }

  + .btn-row {
    margin-top: 5px;
  }
}
