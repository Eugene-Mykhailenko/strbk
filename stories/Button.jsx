import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import { text, select, boolean } from '@storybook/addon-knobs';

import Button from '../src/components/Button/Button';

import styles from './Stories.scss';

storiesOf('Button', module)
  .add(
    'Button',
    withInfo({
      text: `Button component to follow style guide. See the usage below`,
      inline: true,
    })(() => (
      <div className={styles.container}>
        <div className={styles.col}>
          <Button
            text={text('Button text:', 'Button text')}
            intent={select('Style variant:', [
              'main',
              'primary',
              'secondary',
              'cancel',
              'default',
            ])}
            textColor={text('Text color:', '')}
            size={select('Button size:', ['md', 'sm'])}
            block={boolean('block Button:', false)}
            isLoading={boolean('isLoading Button:', false)}
            disabled={boolean('disabled Button:', false)}
            icon={text('Icon name:', '')}
            iconSize={text('Icon Size:')}
            onClick={action('On click handler')}
          />
        </div>
      </div>
    )),
  )
  .add(
    'Buttons examples',
    withInfo({
      inline: true,
      source: false,
    })(() => (
      <div className={styles.container}>
        <div className={styles.col}>
          <h2 className={styles.title}>Button Icon.</h2>
          <div className={Button.styles['btn-row']}>
            <Button intent="simple" icon="UII_BurgerMenu" />
          </div>

          <h2 className={styles.title}> Button with Icon.</h2>
          <div className={Button.styles['btn-row']}>
            <Button text="Button text" icon="UII_Bonuses" iconSize={17} />
          </div>

          <h2 className={styles.title}>Button small.</h2>
          <div className={Button.styles['btn-row']}>
            <Button
              text="Button text"
              size="sm"
              icon="UII_Bonuses"
              iconSize={15}
            />
          </div>

          <h2 className={styles.title}>Button loading.</h2>
          <div className={Button.styles['btn-row']}>
            <Button text="Button text" isLoading />
          </div>

          <h2 className={styles.title}>Button disabled.</h2>
          <div className={Button.styles['btn-row']}>
            <Button
              text="Button text"
              icon="UII_Bonuses"
              iconSize={17}
              disabled
              isLoading={false}
            />
          </div>

          <h2 className={styles.title}>Buttons Block.</h2>
          <Button text="Button text" intent="main" block />
          <Button text="Button text" intent="primary" block />

          <h2 className={styles.title}>Default buttons in button row </h2>
          <div className={Button.styles['btn-row']}>
            <Button text="Button text" intent="main" />
            <Button text="Button text" intent="primary" />
            <Button text="Button text" intent="secondary" />
            <Button text="Button text" intent="default" />
          </div>

          <h2 className={styles.title}>Block buttons in button row </h2>
          <div className={Button.styles['btn-row']}>
            <Button text="Button text" intent="main" block />
            <Button text="Button text" intent="primary" block />
            <Button text="Button text" intent="secondary" block />
          </div>
        </div>
      </div>
    )),
  );
