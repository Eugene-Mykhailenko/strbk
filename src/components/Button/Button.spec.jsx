import React from 'react';
import Button from './Button';

jest.mock('../Icon/importIcons');

const config = {
  text: `Hi it's a button!`,
  isLoading: true,
  onClick: jest.fn(),
  icon: 'UII_BurgerMenu',
};

describe('Button', () => {
  beforeEach(() => {
    config.onClick.mockClear();
  });

  it('should handle click event', () => {
    const button = mount(<Button onClick={config.onClick} />);

    button.simulate('click');

    expect(config.onClick).toHaveBeenCalledTimes(1);

    config.onClick.mockClear();

    expect(button).toMatchSnapshot();
  });

  it('should be of button type by default', () => {
    const button = shallow(<Button onClick={config.onClick} />);

    expect(button.props().type).toBe('button');
    expect(button).toMatchSnapshot();
  });

  it('should be able to indicate loading state', () => {
    const button = shallow(
      <Button isLoading={config.isLoading} onClick={config.onClick} />,
    );

    expect(button.prop('className')).toMatch(/is-loading/);
  });

  it('should render provider text', () => {
    const button = shallow(
      <Button text={config.text} onClick={config.onClick} />,
    );

    expect(button.text()).toEqual(config.text);
  });

  it('should be able to render provider icon', () => {
    const button = shallow(
      <Button icon={config.icon} onClick={config.onClick} />,
    );

    expect(button.childAt(0).props().glyph).toEqual(config.icon);
    expect(button).toMatchSnapshot();
  });

  it('should be able to render in a disabled state', () => {
    const isDisabled = true;

    const button = shallow(
      <Button disabled={isDisabled} onClick={config.onClick} />,
    );

    expect(button.prop('disabled')).toBe(true);
    expect(button).toMatchSnapshot();
  });

  it('should not handle click in a disabled state', () => {
    const isDisabled = true;

    const button = mount(
      <Button disabled={isDisabled} onClick={config.onClick} />,
    );

    button.simulate('click');

    expect(config.onClick).toHaveBeenCalledTimes(0);

    config.onClick.mockClear();

    expect(button).toMatchSnapshot();
  });
});
