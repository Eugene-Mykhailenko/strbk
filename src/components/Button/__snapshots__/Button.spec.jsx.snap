// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be able to render in a disabled state 1`] = `
<button
  className="btn btn-main disabled"
  disabled={true}
  onClick={[Function]}
  type="button"
/>
`;

exports[`Button should be able to render provider icon 1`] = `
<button
  className="btn btn-main"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <Icon
    className="btn-icon"
    fill="currentColor"
    glyph="UII_BurgerMenu"
    height={20}
    width={20}
  />
</button>
`;

exports[`Button should be of button type by default 1`] = `
<button
  className="btn btn-main"
  disabled={false}
  onClick={[Function]}
  type="button"
/>
`;

exports[`Button should handle click event 1`] = `
<Button
  block={false}
  disabled={false}
  icon={null}
  iconSize={20}
  intent="main"
  isLoading={false}
  onClick={[MockFunction]}
  type="button"
>
  <button
    className="btn btn-main"
    disabled={false}
    onClick={[Function]}
    type="button"
  />
</Button>
`;

exports[`Button should not handle click in a disabled state 1`] = `
<Button
  block={false}
  disabled={true}
  icon={null}
  iconSize={20}
  intent="main"
  isLoading={false}
  onClick={[MockFunction]}
  type="button"
>
  <button
    className="btn btn-main disabled"
    disabled={true}
    onClick={[Function]}
    type="button"
  />
</Button>
`;
